contracts/core/Registry.sol

```diff
@@ -1,47 +1,242 @@
+    struct Identity {
+        uint256 nonce;
+        address owner;
         string name;
-        string metadata; // @todo update to MetaPtr
-        address attestationAddr;
+        Metadata metadata;
+        address anchor;
     }
 
-    // getter for projects mapping
-    function identities(uint _identityId) external view returns (IdentityDetails memory) {
+    /// ==========================
+    /// === Storage Variables ====
+    /// ==========================
+
+    /// @notice Identity.id -> Identity
+    mapping(bytes32 => Identity) public identitiesById;
+
+    /// @notice anchor -> Identity.id
+    mapping(address => bytes32) public anchorToIdentityId;
+
+    /// ======================
+    /// ======= Events =======
+    /// ======================
+
+    event IdentityCreated(
+        bytes32 indexed identityId, uint256 nonce, string name, Metadata metadata, address owner, address anchor
+    );
+    event IdentityNameUpdated(bytes32 indexed identityId, string name, address anchor);
+    event IdentityMetadataUpdated(bytes32 indexed identityId, Metadata metadata);
+    event IdentityOwnerUpdated(bytes32 indexed identityId, address owner);
+
+    /// ====================================
+    /// =========== Modifier ===============
+    /// ====================================
+
+    modifier isIdentityOwner(bytes32 _identityId) {
+        if (!isOwnerOfIdentity(_identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+        _;
+    }
+
+    /// ====================================
+    /// ==== External/Public Functions =====
+    /// ====================================
+
+    /// @notice Retrieve identity by identityId
+    /// @param identityId The identityId of the identity
+    function getIdentityById(bytes32 identityId) public view returns (Identity memory) {
+        return identitiesById[identityId];
+    }
+
+    /// @notice Retrieve identity by anchor
+    /// @param anchor The anchor of the identity
+    function getIdentityByAnchor(address anchor) public view returns (Identity memory) {
+        bytes32 identityId = anchorToIdentityId[anchor];
+        return identitiesById[identityId];
+    }
+
+    /// @notice Creates a new identity
+    /// @dev This will also set the attestation address generated from msg.sender and name
+    /// @param _nonce Nonce used to generate identityId
+    /// @param _name The name of the identity
+    /// @param _metadata The metadata of the identity
+    /// @param _members The members of the identity
+    /// @param _owner The owner of the identity
+    function createIdentity(
+        uint256 _nonce,
+        string memory _name,
+        Metadata memory _metadata,
+        address _owner,
+        address[] memory _members
+    ) external returns (bytes32) {
+        bytes32 identityId = _generateIdentityId(_nonce);
+
+        if (identitiesById[identityId].nonce != 0) {
+            revert NONCE_NOT_AVAILABLE();
+        }
```

Comment:
Careful using nonce here. What if they use a `nonce` of 0? Then they'd be able to overwrite it.

I'd recommend explicitly checking that `_owner != address(0)` when this is called, and then reverting if `identitiesById[identityId].owner != address(0)`

Reply:
Ah good callout

Reply:
I would add an explicit check that `_owner != address(0)` so you are sure this invariant is rock solid.

If not, there's the possibility someone could choose address(0) as an owner (maybe wanting to make it immutable) but not realize that would allow accidentally using the same nonce again and overwriting.

Reply:
Actually, you know what? Doing `anchor` for this is better because it'll never be zero, since nobody knows the preimage.